File: api/agents/pricing_analyst_agent.py
"""
Pricing Analyst Agent - Generates data-driven pricing proposals
"""
from api.agents.base import BaseAgent
from api.database import get_session, ListingRepository, CalendarRepository, ProposalRepository
from api.models.proposal import Proposal
from datetime import date, datetime, timedelta
from decimal import Decimal
from typing import Dict, List
import logging

logger = logging.getLogger(__name__)


class PricingAnalystAgent(BaseAgent):
    """
    Agent for generating pricing proposals

    Responsibilities:
    - Analyze occupancy rates and market signals
    - Generate data-driven pricing proposals
    - Classify risk level (low/medium/high)
    - Apply floor/ceiling price constraints
    """

    def __init__(self, api_key: str, env: str = "prod"):
        super().__init__(api_key, env)

        # Create lyzr-adk agent
        self.agent = self.studio.create_agent(
            name="Pricing Analyst",
            provider="openai/gpt-4o",
            role="Revenue management pricing analyst",
            goal="Generate data-driven pricing proposals for Dubai properties",
            instructions="""
            You are a pricing analyst for short-term rental properties in Dubai.

            Your responsibilities:
            - Analyze occupancy rates, events, and market signals
            - Generate pricing proposals with clear reasoning
            - Classify risk level:
              * LOW: 0-10% price change, supported by strong signals
              * MEDIUM: 11-20% price change, moderate signals
              * HIGH: >20% price change, requires strong justification
            - Always respect floor/ceiling price constraints
            - Group consecutive days with similar signals
            - Provide clear, data-driven reasoning

            Never suggest prices below floor or above ceiling.
            Always explain why a price change is recommended.
            """,
            temperature=0.5,
            memory=30,  # Remember context across proposals
        )

        # Add tools
        self.agent.add_tool(self.get_listing_data)
        self.agent.add_tool(self.calculate_occupancy)
        self.agent.add_tool(self.get_calendar_prices)
        self.agent.add_tool(self.save_proposal)

        logger.info("Pricing Analyst Agent initialized")

    def get_listing_data(self, listing_id: int) -> Dict:
        """
        Get listing details for pricing context

        Args:
            listing_id: Listing ID

        Returns:
            Listing data
        """
        try:
            with next(get_session()) as session:
                repo = ListingRepository(session)
                listing = repo.get_by_id(listing_id)

                if not listing:
                    return {"error": f"Listing {listing_id} not found"}

                return {
                    "id": listing.id,
                    "name": listing.name,
                    "current_price": float(listing.price),
                    "price_floor": float(listing.price_floor),
                    "price_ceiling": float(listing.price_ceiling),
                    "bedrooms": listing.bedrooms_number,
                    "city": listing.city,
                    "area": listing.area,
                }
        except Exception as e:
            logger.error(f"Error getting listing data: {e}")
            return {"error": str(e)}

    def calculate_occupancy(self, listing_id: int, days: int = 30) -> Dict:
        """
        Calculate occupancy rate for last N days

        Args:
            listing_id: Listing ID
            days: Number of days to analyze

        Returns:
            Occupancy statistics
        """
        try:
            with next(get_session()) as session:
                repo = CalendarRepository(session)
                end_date = date.today()
                start_date = end_date - timedelta(days=days)

                calendar = repo.get_by_listing_and_range(
                    listing_id, start_date, end_date
                )

                if not calendar:
                    return {"occupancy_rate": 0, "days_analyzed": 0, "message": "No calendar data"}

                booked_days = sum(1 for day in calendar if day.status == "booked")
                total_days = len(calendar)
                occupancy_rate = round((booked_days / total_days) * 100, 2)

                return {
                    "occupancy_rate": occupancy_rate,
                    "booked_days": booked_days,
                    "total_days": total_days,
                    "period": f"{days} days",
                    "message": f"{occupancy_rate}% occupancy over last {days} days"
                }
        except Exception as e:
            logger.error(f"Error calculating occupancy: {e}")
            return {"error": str(e)}

    def get_calendar_prices(self, listing_id: int, start_date: str, end_date: str) -> List[Dict]:
        """
        Get current prices for date range

        Args:
            listing_id: Listing ID
            start_date: Start date (YYYY-MM-DD)
            end_date: End date (YYYY-MM-DD)

        Returns:
            List of calendar prices
        """
        try:
            with next(get_session()) as session:
                repo = CalendarRepository(session)
                calendar = repo.get_by_listing_and_range(
                    listing_id,
                    datetime.strptime(start_date, "%Y-%m-%d").date(),
                    datetime.strptime(end_date, "%Y-%m-%d").date()
                )

                return [
                    {
                        "date": day.date.isoformat(),
                        "price": float(day.price),
                        "status": day.status,
                        "minimum_stay": day.minimum_stay,
                    }
                    for day in calendar
                ]
        except Exception as e:
            logger.error(f"Error getting calendar prices: {e}")
            return []

    def save_proposal(self, proposal_data: Dict) -> Dict:
        """
        Save pricing proposal to database

        Args:
            proposal_data: Proposal details

        Returns:
            Saved proposal info
        """
        try:
            with next(get_session()) as session:
                repo = ProposalRepository(session)

                proposal = Proposal(
                    listing_id=proposal_data["listing_id"],
                    date_range_start=datetime.strptime(proposal_data["date_range_start"], "%Y-%m-%d").date(),
                    date_range_end=datetime.strptime(proposal_data["date_range_end"], "%Y-%m-%d").date(),
                    current_price=Decimal(str(proposal_data["current_price"])),
                    proposed_price=Decimal(str(proposal_data["proposed_price"])),
                    change_pct=proposal_data["change_pct"],
                    risk_level=proposal_data["risk_level"],
                    reasoning=proposal_data["reasoning"],
                    signals=proposal_data.get("signals", {})
                )

                saved = repo.create(proposal)

                return {
                    "proposal_id": saved.id,
                    "status": "saved",
                    "message": f"Proposal {saved.id} created successfully"
                }
        except Exception as e:
            logger.error(f"Error saving proposal: {e}")
            return {"error": str(e)}

    def generate_proposals(self, listing_id: int, start_date: str, end_date: str) -> Dict:
        """
        Generate pricing proposals for date range

        Args:
            listing_id: Listing ID
            start_date: Start date (YYYY-MM-DD)
            end_date: End date (YYYY-MM-DD)

        Returns:
            Generation result
        """
        try:
            response = self.agent.run(
                f"""Generate pricing proposals for listing {listing_id} from {start_date} to {end_date}.

                Steps:
                1. Get listing data and price constraints (floor/ceiling)
                2. Calculate recent occupancy rate
                3. Get current calendar prices
                4. Analyze signals (occupancy, events, seasonality)
                5. Generate proposals respecting floor/ceiling constraints
                6. Classify risk level based on price change percentage
                7. Save proposals to database

                Return summary of proposals created with reasoning.""",
                stream=False
            )

            return {
                "success": True,
                "listing_id": listing_id,
                "date_range": f"{start_date} to {end_date}",
                "response": response.response,
                "status": "completed"
            }
        except Exception as e:
            logger.error(f"Error generating proposals: {e}")
            return {
                "success": False,
                "listing_id": listing_id,
                "error": str(e),
                "status": "failed"
            }

File: src/lib/agents/pricing-analyst-agent.ts
import { db, inventoryMaster, listings, marketEvents } from "@/lib/db";
import { eq, and, gte, lte, sql } from "drizzle-orm";
import { format, addDays, eachDayOfInterval, parseISO, differenceInDays } from "date-fns";
import { createEventIntelligenceAgent } from "./event-intelligence-agent";
import type { EventSignal } from "./event-intelligence-agent";

export interface PricingProposal {
  listingId: number;
  date: string; // ISO date YYYY-MM-DD
  currentPrice: number;
  proposedPrice: number;
  priceFloor: number;
  priceCeiling: number;
  changePct: number;
  riskLevel: "low" | "medium" | "high";
  reasoning: string;
}

export interface AnalysisResult {
  proposals: PricingProposal[];
  summary: string;
  totalProposals: number;
  averageIncrease: number;
}

/**
 * Pricing Analyst Agent
 * Responsible for generating pricing proposals based on data and signals
 */
export class PricingAnalystAgent {
  private eventAgent = createEventIntelligenceAgent();

  /**
   * Generate proposals for a specific listing and date range
   */
  async generateProposals(
    listingId: number,
    startDate: Date,
    endDate: Date
  ): Promise<AnalysisResult> {
    const proposals: PricingProposal[] = [];

    // Fetch listing details
    const [listing] = await db
      .select()
      .from(listings)
      .where(eq(listings.id, listingId))
      .limit(1);

    if (!listing) {
      throw new Error(`Listing ${listingId} not found`);
    }

    // Fetch events for date range
    const eventAnalysis = await this.eventAgent.analyzeEvents(startDate, endDate);

    // Fetch current calendar prices
    const calendar = await db
      .select()
      .from(inventoryMaster)
      .where(
        and(
          eq(inventoryMaster.listingId, listingId),
          gte(inventoryMaster.date, format(startDate, "yyyy-MM-dd")),
          lte(inventoryMaster.date, format(endDate, "yyyy-MM-dd"))
        )
      );

    // Calculate occupancy rate (last 30 days)
    const occupancy = await this.calculateOccupancy(listingId);

    // Generate proposals for each day
    const days = eachDayOfInterval({ start: startDate, end: endDate });

    for (const day of days) {
      const dateStr = format(day, "yyyy-MM-dd");

      // Get current price from calendar or fallback to base
      const calendarDay = calendar.find(d => d.date === dateStr);
      const currentPrice = calendarDay ? parseFloat(calendarDay.currentPrice) : parseFloat(listing.price);

      // Dynamic floor/ceiling (since removed from listings table)
      // Default: Floor = 50% of base, Ceiling = 300% of base
      const basePrice = parseFloat(listing.price);
      const priceFloor = Math.round(basePrice * 0.5);
      const priceCeiling = Math.round(basePrice * 3.0);

      // Check for events on this day
      const dayEvents = eventAnalysis.events.filter(e =>
        e.startDate <= dateStr && e.endDate >= dateStr
      );

      // Calculate proposed price based on signals
      let proposedPrice = currentPrice;
      let reasoning = "";

      // Event-based pricing
      if (dayEvents.length > 0) {
        const recommendation = this.eventAgent.getPricingRecommendation(dayEvents);
        const increase = (currentPrice * recommendation.suggestedIncrease) / 100;
        proposedPrice = currentPrice + increase;
        reasoning = recommendation.reasoning;
      } else {
        // Base pricing on occupancy
        if (occupancy > 80) {
          proposedPrice = currentPrice * 1.1; // 10% increase for high occupancy
          reasoning = `High occupancy (${occupancy}%). Demand is strong, increase pricing.`;
        } else if (occupancy < 60) {
          proposedPrice = currentPrice * 0.95; // 5% decrease for low occupancy
          reasoning = `Low occupancy (${occupancy}%). Decrease price to attract bookings.`;
        } else {
          proposedPrice = currentPrice;
          reasoning = `Moderate occupancy (${occupancy}%). Maintain current pricing.`;
        }
      }

      // Apply floor/ceiling constraints
      if (proposedPrice < priceFloor) {
        proposedPrice = priceFloor;
        reasoning += ` (Capped at floor: AED ${priceFloor})`;
      } else if (proposedPrice > priceCeiling) {
        proposedPrice = priceCeiling;
        reasoning += ` (Capped at ceiling: AED ${priceCeiling})`;
      }

      // Round to nearest 10
      proposedPrice = Math.round(proposedPrice / 10) * 10;

      const changePct = Math.round(((proposedPrice - currentPrice) / currentPrice) * 100);

      // Skip if no change
      if (Math.abs(changePct) < 1) {
        continue;
      }

      // Determine risk level
      // Determine risk level
      const riskLevel = this.calculateRiskLevel(changePct, dayEvents.length);

      proposals.push({
        listingId,
        date: dateStr,
        currentPrice,
        proposedPrice,
        priceFloor,
        priceCeiling,
        changePct,
        riskLevel,
        reasoning,
      });
    }

    // Calculate summary stats
    const totalProposals = proposals.length;
    const averageIncrease =
      proposals.reduce((sum, p) => sum + p.changePct, 0) / (totalProposals || 1);

    let summary = "";
    if (totalProposals === 0) {
      summary = "No pricing changes recommended for this period.";
    } else {
      summary = `${totalProposals} proposal(s) generated with avg ${averageIncrease > 0 ? "+" : ""}${averageIncrease.toFixed(1)}% change.`;
    }

    return {
      proposals,
      summary,
      totalProposals,
      averageIncrease,
    };
  }

  /**
   * Save proposals to database
   */
  async saveProposals(analysisResult: AnalysisResult): Promise<number[]> {
    const proposalIds: number[] = [];

    for (const proposal of analysisResult.proposals) {
      const [inserted] = await db
        .update(inventoryMaster)
        .set({
          proposedPrice: proposal.proposedPrice.toString(),
          changePct: proposal.changePct,
          proposalStatus: "pending",
          reasoning: proposal.reasoning,
        })
        .where(
          and(
            eq(inventoryMaster.listingId, proposal.listingId),
            eq(inventoryMaster.date, proposal.date)
          )
        )
        // @ts-ignore
        .returning({ id: inventoryMaster.id });

      if (inserted) {
        proposalIds.push(inserted.id);
      }
    }

    return proposalIds;
  }

  /**
   * Calculate occupancy rate for last 30 days
   */
  private async calculateOccupancy(listingId: number): Promise<number> {
    const thirtyDaysAgo = addDays(new Date(), -30);
    const today = new Date();

    const calendar = await db
      .select()
      .from(inventoryMaster)
      .where(
        and(
          eq(inventoryMaster.listingId, listingId),
          gte(inventoryMaster.date, format(thirtyDaysAgo, "yyyy-MM-dd")),
          lte(inventoryMaster.date, format(today, "yyyy-MM-dd"))
        )
      );

    if (calendar.length === 0) return 0;

    const bookedDays = calendar.filter((day) => day.status === "booked").length;
    return Math.round((bookedDays / calendar.length) * 100);
  }

  /**
   * Group consecutive dates by event impact
   */
  private groupDatesByEventImpact(
    dates: Date[],
    events: EventSignal[]
  ): Array<{ start: Date; end: Date; events: EventSignal[] }> {
    const groups: Array<{ start: Date; end: Date; events: EventSignal[] }> = [];

    let currentGroup: { start: Date; end: Date; events: EventSignal[] } | null = null;

    for (const date of dates) {
      const dateStr = format(date, "yyyy-MM-dd");

      // Find events overlapping this date
      const overlappingEvents = events.filter((event) => {
        return dateStr >= event.startDate && dateStr <= event.endDate;
      });

      // Check if event impact changed
      const impactChanged =
        currentGroup &&
        (overlappingEvents.length !== currentGroup.events.length ||
          !overlappingEvents.every((e) =>
            currentGroup!.events.some((ce) => ce.name === e.name)
          ));

      if (!currentGroup || impactChanged) {
        // Start new group
        if (currentGroup) {
          groups.push(currentGroup);
        }
        currentGroup = {
          start: date,
          end: date,
          events: overlappingEvents,
        };
      } else {
        // Extend current group
        currentGroup.end = date;
      }
    }

    if (currentGroup) {
      groups.push(currentGroup);
    }

    return groups;
  }

  /**
   * Calculate risk level based on price change and event confidence
   */
  private calculateRiskLevel(
    changePct: number,
    eventCount: number
  ): "low" | "medium" | "high" {
    const absChange = Math.abs(changePct);

    if (absChange <= 10 && eventCount > 0) {
      return "low"; // Small change with event backing
    }

    if (absChange <= 30 && eventCount > 0) {
      return "medium"; // Moderate change with event backing
    }

    if (absChange > 30) {
      return "high"; // Large change, risky
    }

    if (eventCount === 0 && absChange > 10) {
      return "medium"; // No event backing, moderate change
    }

    return "low";
  }

  /**
   * Generate proposals for all properties
   */
  async generatePortfolioProposals(
    startDate: Date,
    endDate: Date
  ): Promise<Map<number, AnalysisResult>> {
    const allListings = await db.select().from(listings);
    const results = new Map<number, AnalysisResult>();

    for (const listing of allListings) {
      const result = await this.generateProposals(listing.id, startDate, endDate);
      if (result.totalProposals > 0) {
        results.set(listing.id, result);
      }
    }

    return results;
  }
}

/**
 * Create a Pricing Analyst Agent instance
 */
export function createPricingAnalystAgent(): PricingAnalystAgent {
  return new PricingAnalystAgent();
}
